### PR TITLE
Add feature for maximum submissions per month

### LINKED ISSUE #1690: Add support for setting submission limit  per month per team for a challenge
## Current Status

Currently, we only provide the capability to set the daily submission limit and total submission limit throughout the challenge. For evaluation of tasks envisioning evaluation of models on EvalAI as long-running test server and not challenge with finite time, we would like to add the feature to limit the number of submissions to a challenge per month per team which eventually limits the number of submissions to test-set per publication.

Thanks @jcjohnson for the feature request.

## Deliverables

- [ ] Add new field `max_submission_per_month` for each challenge phase
- [ ] Set default value to `max_submission_in_total`
- [ ] Add check for monthly submission limit 
- [ ] Add tests
- [ ] Update challenge creation API
- [ ] Modify API to get the remaining submissions and add support to show remaining monthly submissions
- [ ] Update yaml documentation
- [ ] Update EvalAI-Starters to add support for monthly max submission limit

### PR DESCRIPTION
**Pull request by RishabhJain2018**
Fix #1690 

Thanks, @vkartik97 for building this feature.
There were few modifications needed and DateTime warnings also need to be fixed so I added them.

### CODE REVIEW

File: apps/jobs/views.py [hunk 1]
```diff
@@ -417,65 +417,76 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
...
+    elif submissions_done_this_month_count >= max_submissions_per_month_count:
+        date_time_now = timezone.now()
+        next_month_date_time = date_time_now + datetime.timedelta(days=+30)
+        next_month_start_date = next_month_date_time.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
+        remaining_time = next_month_start_date - date_time_now
+
+        if submissions_done_today_count >= max_submissions_per_day_count:
+            response_data = {'message': 'Both daily and monthly submission limits are exhausted!',
+                             'remaining_time': remaining_time}
+        else:
```
Comment (deshraj) [inline apps/jobs/views.py]: Test case missing for this condition.

File: apps/jobs/views.py [hunk 2]
```diff
@@ -417,65 +417,76 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
...
         # Get the midnight time of the day i.e. 12:00 AM of next day.
-        midnight = utc.localize(datetime.datetime.combine(
-            date_time_tomorrow, datetime.time()))
-        # Subtract the current time from the midnight time to get the remaining time for the next day's submissions.
+        midnight = date_time_tomorrow.replace(hour=0, minute=0, second=0)
         remaining_time = midnight - date_time_now
-        # Return the remaining time with a message.
-        response_data = {'message': 'You have exhausted today\'s submission limit',
+
+        response_data = {'message': 'You have exhausted today\'s submission limit!',
```
Comment (deshraj) [inline apps/jobs/views.py]: Please indent this code properly.

File: apps/jobs/models.py [hunk 3]
```diff
@@ -133,25 +133,30 @@ def save(self, *args, **kwargs):
...
-                    participant_team=self.participant_team,
-                    status=Submission.FAILED,
-                    # Get the midnight time of the day
-                    submitted_at__gte=timezone.now().replace(hour=0, minute=0, second=0, microsecond=0)).count()
-
-                if ((submissions_done_today_count + 1 - failed_count > self.challenge_phase.max_submissions_per_day) or
-                        (self.challenge_phase.max_submissions_per_day == 0)):
-                    logger.info("Permission Denied: The maximum number of submission for today has been reached")
-                    raise PermissionDenied({'error': 'The maximum number of submission for today has been reached'})
+            submissions_done_today = Submission.objects.filter(
```
Comment (deshraj) [inline apps/jobs/models.py]: You can be smart about making queries to the database. First, pick the common filter attributes and store them in a query set. Then use that queryset to filter out based on other uncommon attributes.

File: apps/jobs/views.py (hunk 2 above)
Comment (deshraj) [reply apps/jobs/views.py]: Line 470-472 is not indented properly. Please see how other dicts are indented.

File: apps/jobs/views.py [hunk 4]
```diff
@@ -417,65 +417,73 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
...
-
-        if remaining_submissions_today_count > remaining_submission_count:
-            remaining_submissions_today_count = remaining_submission_count
-
-        # Return the above calculated data.
-        response_data = {'remaining_submissions_today_count': remaining_submissions_today_count,
-                         'remaining_submissions': remaining_submission_count
-                         }
+                                             submissions_done_today_count)
+        response_data = {'remaining_submissions_this_month_count': remaining_submissions_this_month_count,
```
Comment (deshraj) [inline apps/jobs/views.py]: Please indent properly. Something like this:

```
response_data = {
    'remaining_submissions_this_month_count': remaining_submissions_this_month_count,
    'remaining_submissions_today_count': remaining_submissions_today_count,
      'remaining_submissions': remaining_submission_count,
}

File: tests/unit/jobs/test_views.py [hunk 5]
```diff
@@ -568,47 +605,68 @@ def test_get_remaining_submission_when__submission_made_today(self):
...
+
+        self.challenge.participant_teams.add(self.participant_team)
+        self.challenge.save()
+        response = self.client.get(self.url, {})
+        self.assertEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def get_remaining_submission_time_when_monthly_limit_is_exhausted(self):
+        self.url = reverse_lazy('jobs:get_remaining_submissions',
+                                kwargs={'challenge_phase_pk': self.challenge_phase.pk,
```
Comment (deshraj) [inline tests/unit/jobs/test_views.py]: Please indent it properly. 🙏

File: tests/unit/jobs/test_views.py [hunk 6]
```diff
@@ -568,47 +605,68 @@ def test_get_remaining_submission_when__submission_made_today(self):
...
-        submission_count = self.challenge_phase.max_submissions - other_submissions - failed_submissions
+        self.challenge.participant_teams.add(self.participant_team)
+        self.challenge.save()
+        response = self.client.get(self.url, {})
+        self.assertEqual(response.data['message'], expected['message'])
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def get_remaining_submission_time_when_both_monthly_and_daily_limit_is_exhausted(self):
+        self.url = reverse_lazy('jobs:get_remaining_submissions',
+                                kwargs={'challenge_phase_pk': self.challenge_phase.pk,
```
Comment (deshraj) [inline tests/unit/jobs/test_views.py]: Please indent properly.

**Comment by RishabhJain2018** [comment]
Closing this since #2005 is opened for it.

**Pull request closed without merging**